insights/C/generate/generate_linux.go

```diff
@@ -0,0 +1,70 @@
+//go:build tools
+
+package main
+
+import (
+	"fmt"
+	"log/slog"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+
+	"github.com/ubuntu/ubuntu-insights/insights/internal/constants"
+)
+
+func main() {
+	if err := buildSharedLibs(); err != nil {
+		os.Exit(1)
+	}
+
+	if err := copyTypesHeader(); err != nil {
+		os.Exit(1)
+	}
+}
+
+func buildSharedLibs() error {
+	if version := os.Getenv("DEB_VERSION_UPSTREAM"); version != "" {
+		constants.Version = version
+	}
+
+	re := regexp.MustCompile(`(\d+)\.(\d+)\.(\d+)`)
+	matches := re.FindStringSubmatch(constants.Version)
+
+	libname := "libinsights.so"
+	if len(matches) == 4 {
+		libname = fmt.Sprintf("libinsights.so.%s", matches[1])
+	}
+
+	ldflags := fmt.Sprintf("-X=constants.Version/internal/constants.Version=%s -extldflags -Wl,-soname,%s", constants.Version, libname)
+
+	cmd := exec.Command("go", "build", //nolint:gosec // This is controlled by the build process and also filtered here
+		"-buildmode=c-shared",
+		"-trimpath",
+		"-ldflags", ldflags,
+		"-o", fmt.Sprintf("../generated/%s", libname),
+		"./libinsights.go",
+	)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		slog.Error("Failed to build shared libraries", "stdout", string(output), "error", err)
+		return err
+	}
+
+	// Rename the header file to insights.h
+	lastDot := strings.LastIndex(libname, ".")
+	expectedHeader := libname[:lastDot] + ".h"
+	if err := os.Rename(fmt.Sprintf("../generated/%s", expectedHeader), "../generated/insights.h"); err != nil {
+		slog.Error("Failed to rename header file", "error", err)
+	}
+	return err
+}
+
+func copyTypesHeader() error {
+	cmd := exec.Command("cp", "./insights_types.h", "../generated/types.h")
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		slog.Error("Failed to copy types header", "stdout", string(output), "error", err)
```

Comment:
see above on logging (at least, here, we return the correct error) + combining the lines.